src/AusPost.php

```diff
@@ -4,65 +4,272 @@
 
 namespace ParcelTrap\AusPost;
 
-use DateTimeImmutable;
 use GrahamCampbell\GuzzleFactory\GuzzleFactory;
 use GuzzleHttp\ClientInterface;
+use GuzzleHttp\Exception\ClientException;
 use GuzzleHttp\RequestOptions;
 use ParcelTrap\Contracts\Driver;
 use ParcelTrap\DTOs\TrackingDetails;
 use ParcelTrap\Enums\Status;
+use ParcelTrap\Exceptions\ApiAuthenticationFailedException;
+use ParcelTrap\Exceptions\ApiLimitReachedException;
 
 class AusPost implements Driver
 {
     public const IDENTIFIER = 'auspost';
 
     public const BASE_URI = 'https://digitalapi.auspost.com.au';
 
+    public const API_LIMIT_PERIOD = 'minute';
+
+    public const API_LIMIT_PER_PERIOD = 10;
+
     private ClientInterface $client;
 
-    public function __construct(private readonly string $apiKey, ?ClientInterface $client = null)
+    public function __construct(private readonly string $apiKey, private readonly string $password, private readonly string $accountNumber, ?ClientInterface $client = null)
     {
         $this->client = $client ?? GuzzleFactory::make(['base_uri' => self::BASE_URI]);
     }
 
     public function find(string $identifier, array $parameters = []): TrackingDetails
     {
-        $request = $this->client->request('GET', '/tracking', [
-            RequestOptions::HEADERS => $this->getHeaders(),
-            RequestOptions::QUERY => array_merge(['id' => $identifier], $parameters),
-        ]);
+        // find() only supports returning a single TrackingDetails object so we can only search for one
+        if (str_contains($identifier, ',')) {
+            /** @phpstan-ignore-next-line */
+            $identifier = substr($identifier, 0, strpos($identifier, ','));
+        }
+
+        try {
+            $response = $this->client->request('GET', '/shipping/v1/track', [
+                RequestOptions::HEADERS => $this->getHeaders(),
+                RequestOptions::QUERY => array_merge(['tracking_ids' => $identifier], $parameters),
+            ]);
+        } catch (ClientException $exception) {
+            if ($exception->getCode() === 429) {
+                throw new ApiLimitReachedException(
+                    driver: $this,
+                    limit: static::API_LIMIT_PER_PERIOD,
+                    period: static::API_LIMIT_PERIOD,
+                    previous: $exception,
+                );
+            }
+
+            if (in_array($exception->getCode(), [401, 403])) {
+                throw new ApiAuthenticationFailedException(
+                    driver: $this,
+                    previous: $exception,
+                );
+            }
```

Comment:
Probably want to re-throw the exception here if it's not one of the other types. 👍🏻 That'll sort out the PHPStan warning.

Reply:
Good catch. Bumped up the min version to `^1.2`, added a re-throw, and added a test to ensure the rethrows still work :+1: